File: tool/release.dart
// Cuts a release: bump the version, commit it, tag it, push.
//
// Run with `dart run tool/release.dart`. Pushing the tag is what starts
// .github/workflows/release.yml, which builds the signed APK and creates
// the GitHub release; nothing is built locally.
//
// Written in Dart rather than as a shell script so it behaves the same on
// every machine, and so a Flutter repo does not grow a package.json just
// to hold one script.
import 'dart:io';

const String _pubspecPath = 'pubspec.yaml';

/// `version: 1.0.0+1`, where the part after the plus is Android's
/// versionCode. Play Store and every installer require that number to
/// increase on every build, so it is bumped on every release regardless
/// of which semver part moved.
// [ \t]* rather than \s* at either end: \s matches newlines too, and
// being greedy it would swallow the blank line after the version and
// quietly delete it on every release.
final RegExp _versionLine = RegExp(
  r'^version:[ \t]*(\d+)\.(\d+)\.(\d+)\+(\d+)[ \t]*$',
  multiLine: true,
);

Future<void> main(List<String> args) async {
  _requireRepoRoot();
  await _requireCleanTree();

  final _Version current = _readVersion();
  stdout.writeln('Current version: $current');
  stdout.writeln();

  final _Version next = _promptForNext(current);
  stdout.writeln();
  stdout.writeln('  $current  ->  $next');
  stdout.writeln('  tag: v${next.semver}');
  stdout.writeln();

  if (!_confirm('Release v${next.semver}?')) {
    stdout.writeln('Cancelled. Nothing was changed.');
    exit(1);
  }

  _writeVersion(next);
  stdout.writeln('Updated $_pubspecPath');

  await _run('git', ['add', _pubspecPath]);
  await _run('git', [
    'commit',
    '-m',
    'chore(release): bump version to ${next.semver}',
  ]);
  await _run('git', ['tag', 'v${next.semver}']);
  await _run('git', ['push']);
  await _run('git', ['push', 'origin', 'v${next.semver}']);

  stdout.writeln();
  stdout.writeln('Released v${next.semver}.');
  stdout.writeln(
    'The build runs at https://github.com/symonxdd/aphanes/actions',
  );
}

/// Refuses to run from anywhere but the repository root, since every path
/// here is relative to it.
void _requireRepoRoot() {
  if (!File(_pubspecPath).existsSync()) {
    stderr.writeln('Run this from the repository root.');
    exit(1);
  }
}

/// A dirty tree would get swept into the version commit, which is meant
/// to contain exactly one changed line.
Future<void> _requireCleanTree() async {
  final ProcessResult result = await Process.run('git', [
    'status',
    '--porcelain',
  ]);
  final String output = (result.stdout as String).trim();
  if (output.isNotEmpty) {
    stderr.writeln('Working tree is not clean:');
    stderr.writeln(output);
    stderr.writeln();
    stderr.writeln('Commit or stash first, then release.');
    exit(1);
  }
}

_Version _readVersion() {
  final String pubspec = File(_pubspecPath).readAsStringSync();
  final RegExpMatch? match = _versionLine.firstMatch(pubspec);
  if (match == null) {
    stderr.writeln('No `version: x.y.z+n` line found in $_pubspecPath.');
    exit(1);
  }
  return _Version(
    int.parse(match.group(1)!),
    int.parse(match.group(2)!),
    int.parse(match.group(3)!),
    int.parse(match.group(4)!),
  );
}

void _writeVersion(_Version version) {
  final File file = File(_pubspecPath);
  file.writeAsStringSync(
    file.readAsStringSync().replaceFirst(_versionLine, 'version: $version'),
  );
}

_Version _promptForNext(_Version current) {
  final Map<String, _Version> choices = {
    '1': current.bumpPatch(),
    '2': current.bumpMinor(),
    '3': current.bumpMajor(),
  };
  stdout.writeln('What kind of release is this?');
  stdout.writeln('  1) patch  ${choices['1']!.semver}   a fix, nothing new');
  stdout.writeln('  2) minor  ${choices['2']!.semver}   a new feature');
  stdout.writeln('  3) major  ${choices['3']!.semver}   a breaking change');
  stdout.write('> ');
  final String? answer = stdin.readLineSync()?.trim();
  final _Version? chosen = choices[answer];
  if (chosen == null) {
    stderr.writeln('Not one of the options. Cancelled.');
    exit(1);
  }
  return chosen;
}

bool _confirm(String question) {
  stdout.write('$question [y/N] ');
  final String answer = (stdin.readLineSync() ?? '').trim().toLowerCase();
  return answer == 'y' || answer == 'yes';
}

Future<void> _run(String executable, List<String> arguments) async {
  stdout.writeln('\$ $executable ${arguments.join(' ')}');
  final Process process = await Process.start(
    executable,
    arguments,
    mode: ProcessStartMode.inheritStdio,
  );
  final int code = await process.exitCode;
  if (code != 0) {
    stderr.writeln('Failed with exit code $code. Stopping here.');
    stderr.writeln(
      'Anything already committed or tagged is still in place; undo it '
      'before trying again.',
    );
    exit(code);
  }
}

class _Version {
  const _Version(this.major, this.minor, this.patch, this.build);

  final int major;
  final int minor;
  final int patch;
  final int build;

  String get semver => '$major.$minor.$patch';

  _Version bumpPatch() => _Version(major, minor, patch + 1, build + 1);
  _Version bumpMinor() => _Version(major, minor + 1, 0, build + 1);
  _Version bumpMajor() => _Version(major + 1, 0, 0, build + 1);

  @override
  String toString() => '$semver+$build';
}

File: desktop/scripts/release.mjs
// Makes a desktop release: bump the version, commit it, tag it, push.
//
// Run with `npm run release` from desktop/. Pushing the desktop-vx.y.z tag
// is what starts .github/workflows/release-desktop.yml, which builds the
// signed Windows installer and creates the GitHub release; nothing is
// built locally.
//
// The desktop app is versioned apart from the Android app, which has its
// own release command (tool/release.dart at the repository root), its own
// changelog and plain vx.y.z tags.
import { execFileSync } from "node:child_process";
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import prompts from "prompts";

const CHANGELOG = "CHANGELOG.md";

/**
 * Every place the version lives. They must always agree: tauri.conf.json
 * is the one the installer and the updater read, and the rest are what
 * npm and Cargo read. Each pattern captures the text before the version
 * and the text after it, and is expected to match an exact number of
 * times, so a file whose layout changed stops the release instead of
 * being half updated.
 */
const SITES = [
  { path: "package.json", pattern: /^( {2}"version": ")([^"]+)(")/m, count: 1 },
  { path: "package-lock.json", pattern: /^( {2}"version": ")([^"]+)(")/m, count: 1 },
  { path: "package-lock.json", pattern: /("": \{\s*"name": "aphanes-desktop",\s*"version": ")([^"]+)(")/, count: 1 },
  { path: "src-tauri/tauri.conf.json", pattern: /^( {2}"version": ")([^"]+)(")/m, count: 1 },
  { path: "Cargo.toml", pattern: /^(version = ")([^"]+)(")/m, count: 1 },
  { path: "Cargo.lock", pattern: /(name = "aphanes-(?:desktop|protocol)"\r?\nversion = ")([^"]+)(")/g, count: 2 },
];

function fail(message) {
  console.error(message);
  process.exit(1);
}

function git(...args) {
  console.log(`$ git ${args.join(" ")}`);
  execFileSync("git", args, { stdio: "inherit" });
}

function bump(version, kind) {
  const [major, minor, patch] = version.split(".").map(Number);
  if (kind === "major") return `${major + 1}.0.0`;
  if (kind === "minor") return `${major}.${minor + 1}.0`;
  return `${major}.${minor}.${patch + 1}`;
}

/** Checks every site first, then writes, so a mismatch changes nothing. */
function writeVersion(current, next) {
  const files = new Map();
  for (const { path } of SITES) files.set(path, files.get(path) ?? readFileSync(path, "utf8"));

  for (const { path, pattern, count } of SITES) {
    const global = new RegExp(pattern.source, pattern.flags.includes("g") ? pattern.flags : pattern.flags + "g");
    const found = [...files.get(path).matchAll(global)].map((m) => m[2]);
    if (found.length !== count || found.some((v) => v !== current)) {
      fail(
        `${path}: expected ${count} version entries at ${current}, found ${found.join(", ") || "none"}. Nothing was changed.`,
      );
    }
    files.set(
      path,
      files.get(path).replace(global, (_, before, __, after) => `${before}${next}${after}`),
    );
  }

  for (const [path, text] of files) {
    writeFileSync(path, text);
    console.log(`Updated ${path}`);
  }
  return [...files.keys()];
}

if (!existsSync("src-tauri/tauri.conf.json")) fail("Run this from desktop/, with `npm run release`.");

// A dirty tree would get swept into the version commit, which is meant
// to contain only the version bump.
const dirty = execFileSync("git", ["status", "--porcelain"], { encoding: "utf8" }).trim();
if (dirty) fail(`Working tree is not clean:\n${dirty}\n\nCommit or stash first, then release.`);

const current = JSON.parse(readFileSync("package.json", "utf8")).version;
console.log(`Current desktop version: ${current}\n`);

// The same arrow-key menu as DS4 Dashboard's release script. Escape or
// Ctrl+C leaves it with nothing chosen, which cancels.
const { kind } = await prompts({
  type: "select",
  name: "kind",
  message: "What type of release is this?",
  choices: [
    { title: "Patch", description: `${current} -> ${bump(current, "patch")}, a fix`, value: "patch" },
    { title: "Minor", description: `${current} -> ${bump(current, "minor")}, a new feature`, value: "minor" },
    { title: "Major", description: `${current} -> ${bump(current, "major")}, a breaking change`, value: "major" },
  ],
});
if (!kind) fail("Release cancelled. Nothing was changed.");
const next = bump(current, kind);
const tag = `desktop-v${next}`;

// The release notes are this version's own section of the changelog, so
// without one the workflow would publish a release saying nothing. It
// checks this too, but stopping here costs a moment instead of a build.
const heading = `## v${next}`;
if (
  !readFileSync(CHANGELOG, "utf8")
    .split(/\r?\n/)
    .some((line) => line.trimEnd() === heading)
) {
  fail(`${CHANGELOG} has no "${heading}" section.\nWrite this version's entry before releasing it.`);
}

const changed = writeVersion(current, next);
git("add", ...changed);
git("commit", "-m", `chore(release): bump desktop to ${next}`);
git("tag", tag);
git("push");
git("push", "origin", tag);

console.log(`\nReleased ${tag}.`);
console.log("The build runs at https://github.com/symonxdd/aphanes/actions");
